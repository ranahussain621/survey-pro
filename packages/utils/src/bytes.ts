import { isEmpty } from './helper'

const BYTE = 1
const KB = BYTE * 1024
const MB = KB * 1024
const GB = MB * 1024
const TB = GB * 1024
const PB = TB * 1024
const regx = /^(-?(?:\d+)?\.?\d+)(b|kb|mb|gb|tb|pb)$/i

export function parseBytes(value: string | number): number | undefined {
  if (isEmpty(value)) {
    return
  }

  const str = String(value)
  const matches = str.match(regx)

  if (!matches) {
    return
  }

  const num = parseFloat(matches[1])
  const type = matches[2].toLowerCase()

  switch (type) {
    case 'pb':
      return num * PB

    case 'tb':
      return num * TB

    case 'gb':
      return num * GB

    case 'mb':
      return num * MB

    case 'kb':
      return num * KB

    case 'b':
      return num * BYTE
  }
}

export function bytes(size: string): number | undefined {
  return parseBytes(size)
}

export function formatBytes(value: string | number): string {
  const mag = Math.abs(Number(value))
  let unit: string
  let size: number | string

  if (mag >= PB) {
    unit = 'PB'
    size = mag / PB
  } else if (mag >= TB) {
    unit = 'TB'
    size = mag / TB
  } else if (mag >= GB) {
    unit = 'GB'
    size = mag / GB
  } else if (mag >= MB) {
    unit = 'MB'
    size = mag / MB
  } else if (mag >= KB) {
    unit = 'KB'
    size = mag / KB
  } else {
    unit = 'B'
    size = mag / BYTE
  }

  if (String(size).includes('.')) {
    size = size.toFixed(1)
  }

  return size + unit
}
