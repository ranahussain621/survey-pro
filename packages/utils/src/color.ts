import { isEmpty } from './helper'

const HEX_REGEX = /^#?([0-9A-F]{3}|[0-9A-F]{4}|[0-9A-F]{6}|[0-9A-F]{8})$/i
const RGBA_REGEX =
  /rgba\((\d{1,3}%?),\s*(\d{1,3}%?),\s*(\d{1,3}%?),\s*(\d*(?:\.\d+)?)\)/i
const RGB_REGEX = /rgb\((\d{1,3}%?),\s*(\d{1,3}%?),\s*(\d{1,3}%?)\)/i

export function isHexColor(color: string): boolean {
  return HEX_REGEX.test(color)
}

export function isRgba(color: string): boolean {
  return RGBA_REGEX.test(color)
}

export function isRgb(color: string): boolean {
  return RGB_REGEX.test(color)
}

export function hexToRgb(hex: string): number[] {
  if (isEmpty(hex)) {
    return []
  }

  if (isRgba(hex)) {
    // @ts-ignore
    const [, red, green, blue, alpha] = RGBA_REGEX.exec(hex)
    return [Number(red), Number(green), Number(blue), Number(alpha)]
  }

  if (isRgb(hex)) {
    // @ts-ignore
    const [, red, green, blue] = RGB_REGEX.exec(hex)
    return [Number(red), Number(green), Number(blue), 1]
  }

  if (!isHexColor(hex)) {
    return []
  }

  hex = hex.replace(/^#/, '')

  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2]
  }

  let alpha = 1
  if (hex.length === 8) {
    alpha = hexToFloatAlpha(hex.slice(6, 8))
    hex = hex.slice(0, 6)
  }

  const number = Number.parseInt(hex, 16)
  const red = number >> 16
  const green = (number >> 8) & 255
  const blue = number & 255

  return [red, green, blue, alpha]
}

export function isDarkColor(hex: string): boolean {
  if (!isHexColor(hex)) {
    return true
  }
  return colorBrightness(hex) < 170
}

export function isLightColor(hex: string): boolean {
  return !isDarkColor(hex)
}

export function colorBrightness(hex: string): number {
  const [red, green, blue, alpha] = hexToRgb(hex)
  return (alpha * (red * 299 + green * 587 + blue * 114)) / 1000
}

export function rgbToHex([r, g, b]: number[]): string {
  return '#' + ((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1)
}

export function alpha(hex: string, alpha: number): string {
  const [red, green, blue] = hexToRgb(hex)
  return `rgba(${red}, ${green}, ${blue}, ${alpha})`
}

export function lighten(hex: string, alpha: number): string {
  return rgbToHex(alphaHexToRgb(hex, 1 - alpha, '#fff'))
}

export function darken(hex: string, alpha: number): string {
  return rgbToHex(alphaHexToRgb(hex, 1 - alpha, '#010101'))
}

export function invert(hex: string): string {
  const [red, green, blue] = hexToRgb(hex)
  return rgbToHex([255 - red, 255 - green, 255 - blue])
}

/**
 * Convert rgba to a transparency-adjusted-hex
 * https://stackoverflow.com/a/15898886
 */
export function alphaHexToRgb(
  hex: string,
  alpha: number,
  backgroundHex: string
): number[] {
  const hexRgb = hexToRgb(hex)
  const backgroundRgb = hexToRgb(backgroundHex)

  return hexRgb.map((color, index) => {
    return colorRange(
      Math.floor(
        colorRange(color) * alpha +
          colorRange(backgroundRgb[index]) * (1 - alpha)
      )
    )
  })
}

function colorRange(color: number): number {
  return color < 1 ? 1 : color >= 255 ? 255 : color
}

function hexToFloatAlpha(arg: string) {
  const num = parseInt(arg, 16)
  const max = parseInt('0xff', 16)
  return Number((num / max).toFixed(2))
}
