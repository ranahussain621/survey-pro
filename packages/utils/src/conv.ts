import { isEmpty, isTrue, isBoolean } from './helper'

export function toBool(value: unknown, defaults?: boolean): boolean {
  if (isEmpty(value)) {
    return defaults || false
  }

  if (isBoolean(value)) {
    return value as boolean
  }

  return isTrue(value)
}
