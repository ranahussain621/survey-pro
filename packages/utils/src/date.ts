import { hs } from './hs'
import { OpUnitType, QUnitType } from 'dayjs'
import dayjs from 'dayjs'

export function timestamp(): number {
  return Math.floor(Date.now() / 1e3)
}

export function date(str?: string, format?: string) {
  return dayjs(str, format)
}

export function unixDate(t: number) {
  return dayjs.unix(t)
}

export function isDateExpired(
  start: number,
  end: number,
  expire: string
): boolean {
  return end - start > hs(expire)!
}

export function unixDiff(
  start: number,
  end: number,
  unit: QUnitType | OpUnitType = 'day'
): number {
  if (start > end || start < 0 || end < 0) {
    return 0
  }
  return unixDate(end).diff(unixDate(start), unit)
}

export function datePeriod(
  start: number,
  value = 1,
  unit: OpUnitType = 'month'
): number {
  return unixDate(start)
    .add(value, unit as unknown as any)
    .unix()
}
