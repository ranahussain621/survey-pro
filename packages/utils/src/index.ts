export * from 'deep-object-diff'
export * from './bytes'
export * from './clone'
export * from './color'
export * from './conv'
export * from './date'
export { default as helper } from './helper'
export * from './hs'
export * from './mime'
export * from './nanoid'
export * from './object'
export * from './parse'
export { default as qs } from './qs'
export * from './random'
export * from './slugify'
export * from './type'
export * from './uuid'
