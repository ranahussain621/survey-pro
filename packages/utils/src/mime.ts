import * as m from 'mime-types'

export const commonImageMimeTypes: string[] = [
  m.lookup('.jpg'),
  m.lookup('.png'),
  m.lookup('.bmp'),
  m.lookup('.webp'),
  m.lookup('.gif')
] as any

export const commonFileMimeTypes: string[] = [
  ...commonImageMimeTypes,
  m.lookup('.txt'),
  m.lookup('.md'),
  m.lookup('.doc'),
  m.lookup('.docx'),
  m.lookup('.xls'),
  m.lookup('.xlsx'),
  m.lookup('.csv'),
  m.lookup('.ppt'),
  m.lookup('.pptx'),
  m.lookup('.pdf'),
  m.lookup('.mp4'),
  m.lookup('.wmv'),
  m.lookup('.svg')
] as any

export const mime = m.lookup
