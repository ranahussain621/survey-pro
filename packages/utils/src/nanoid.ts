import { customAlphabet } from 'nanoid'

const NANOID_ALPHABET =
  'ModuleSymbhasOwnPr0123456789ABCDEFGHNRVfgctiUvzKqYTJkLxpZXIjQW'

export function nanoid(len = 21): string {
  return nanoidCustomAlphabet(NANOID_ALPHABET, len)
}

export function nanoidCustomAlphabet(alphabet: string, len = 21): string {
  const generate = customAlphabet(alphabet, len)
  return generate()
}
