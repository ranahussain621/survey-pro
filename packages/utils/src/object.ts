import { clone } from './clone'
import {
  isArray,
  isNan,
  isNil,
  isObject,
  isPlainObject,
  isValid
} from './helper'
import * as objectPath from 'object-path'

export function pickObject(
  target: Record<string, any>,
  fields: Array<string | string[]>,
  excludes: Array<string> = []
): Record<string, any> {
  if (!isObject(target)) {
    return {}
  }

  const copied = clone<Record<string, any>>(target)
  const targetKeys = Object.keys(copied)

  const fieldAlias: Record<string, any> = {}
  let picked: string[] = []

  if (fields.length > 0) {
    fields.forEach(item => {
      if (isArray(item)) {
        if (item.length > 0) {
          const filed = item[0]
          picked.push(filed)

          if (item.length > 1) {
            fieldAlias[filed] = item[1]
          }
        }
      } else {
        picked.push(item)
      }
    })
  } else {
    picked = targetKeys
  }

  const newObj: Record<string, any> = {}

  targetKeys
    .filter(key => !excludes.includes(key) && picked.includes(key))
    .forEach(key => {
      const alias = fieldAlias[key]

      if (alias) {
        newObj[alias] = copied[key]
      } else {
        newObj[key] = copied[key]
      }
    })

  return newObj
}

export function pickValidValues<T = string | number | boolean>(
  target: Record<string, T>,
  fields: Array<string | string[]>
): Record<string, T> {
  const dist: Record<string, T> = {}

  fields.forEach(field => {
    let key = String(field)
    let alias: string | undefined

    if (isArray(field)) {
      key = field[0]

      if (field.length > 1) {
        alias = field[1]
      }
    }

    let value = target[key]

    if (isValid(value)) {
      if (isPlainObject(value) || isArray(value)) {
        value = clone(value)
      }

      dist[alias || key] = value
    }
  })

  return dist
}

export function removeObjectNil(
  target: Record<string, any>
): Record<string, any> {
  if (!isObject(target)) {
    return {}
  }

  const copied = clone<Record<string, any>>(target)
  const newObj: Record<string, any> = {}

  for (const field of Object.keys(copied)) {
    const value = copied[field]

    if (isNil(value) || isNan(value)) {
      continue
    }

    newObj[field] = value
  }

  return newObj
}

export function copyObjectValues(
  target: Record<string, any>,
  dist: Record<string, any>,
  keyMaps: Array<string | string[]>
): void {
  if (!isObject(target) || !isObject(dist)) {
    return
  }

  for (const keys of keyMaps) {
    let targetKey: string
    let distKey: string

    if (isArray(keys)) {
      targetKey = keys[0]
      distKey = keys[1]
    } else {
      targetKey = String(keys)
      distKey = targetKey
    }

    let value = objectPath.get(target, targetKey)

    if (!isNil(value) && !isNan(value)) {
      if (isPlainObject(value) || isArray(value)) {
        value = clone(value)
      }
      objectPath.set(dist, distKey, value)
    }
  }
}
