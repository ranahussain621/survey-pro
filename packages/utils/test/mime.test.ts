import { mime, commonImageMimeTypes, commonFileMimeTypes } from '../src'

test('image mime type', () => {
  expect(mime('.jpg')).toBe('image/jpeg')
})

test('common image mime type', () => {
  expect(commonImageMimeTypes).toMatchSnapshot()
})

test('common file mime type', () => {
  expect(commonFileMimeTypes).toMatchSnapshot()
})
