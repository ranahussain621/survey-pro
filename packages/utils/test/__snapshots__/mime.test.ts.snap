// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common file mime type 1`] = `
Array [
  "image/jpeg",
  "image/png",
  "image/bmp",
  "image/webp",
  "image/gif",
  "text/plain",
  "text/x-markdown",
  "application/msword",
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  "application/vnd.ms-excel",
  "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  "text/csv",
  "application/vnd.ms-powerpoint",
  "application/vnd.openxmlformats-officedocument.presentationml.presentation",
  "application/pdf",
  "video/mp4",
  "video/x-ms-wmv",
  "image/svg+xml",
]
`;

exports[`common image mime type 1`] = `
Array [
  "image/jpeg",
  "image/png",
  "image/bmp",
  "image/webp",
  "image/gif",
]
`;
